### PR TITLE
Add gradient accumulation and optimizer step limits to Brain class

### PR DESCRIPTION
**Pull request by kokamido**
This PR is based on [discourse conversation](https://speechbrain.discourse.group/t/feature-request-iteration-number-in-training-loop/545). The PR provides a possibility to stop training depending on optimizer step count. For some cases (listed in the discourse topic) it's more handy than the epoch counter.
This PR also adds a gradient accumulation inspired by the [CommonVoice ASR transformer recipe](https://github.com/speechbrain/speechbrain/blob/develop/recipes/CommonVoice/ASR/transformer/train.py#L140). It's easier to configure training with optimizer step limit while gradient accumulation is enabled. The existing approach (based on epoch counter) requires calculations every time you want to tweak batch size keeping the same number of optimization steps.
I've also made a branch with tests: https://github.com/kokamido/speechbrain/tree/grad_acc_and_iter_stop_tests
Tests run deterministic training pipeline with various combinations of gradient accumulation factor and optimizer step limit then check that logged validation losses are the same.
You can run the tests as
```
cd speechbrain/tests/grad_acc_test/
sh test_grac_acc.sh
```
Two GPUs are required for testing due to DP/DDP scenarios.
I hope you’ll find this convenient

### CODE REVIEW

**Comment by TParcollet** [comment]
Hey! Thank you so much, these directions look cool and may interest @Gastron and @pplantinga

**Comment by Gastron** [comment]
The looks good to me, but @TParcollet do you think we should update all the recipes that use gradient accumulation to this?

**Comment by mravanelli** [comment]
I'm fine with these changes. 
- As for the gradient accumulation, we already considered adding it in the `fit_batch` function of core.py.  However, as you can see there is often the need to override the `fit_batch` method within the recipe script. For instance, in the [CommonVoice Transformer recipe](https://github.com/speechbrain/speechbrain/blob/develop/recipes/CommonVoice/ASR/transformer/train.py#L150), we need to override it because we have a learning rate annealing done at every update (not supported by the version in `core.py`). 
In this case, we have to re-implement the gradient accumulation again in train.py.

- As it is basically very hard to write a` fit_batch` function that meets everyone's needs, our approach was to keep in `core.py` a basic version of it and leave users all the freedom to customize  ` fit_batch`  based on their needs in train.py. Gradient descends only require a couple of lines of code, so it shouldn't be too hard to add.  Said that I'm fine adding it `core.py` as well if this is considered useful by the users.

So @Gastron, there is not need to update the existing recipes because they already have to override fit_batch.

**Comment by Gastron** [comment]
Right! I think this makes sense to include, and I like how this PR deals with the two different but connected improvements. Merging!

**Comment by RuABraun** [comment]
One slight issue I want to point out is that if one is comparing builds with different amounts of gradient accumulation then in, the current implementation, the loss will be scaled differently making it harder to compare.

The correct thing to do is to sum the loss across the different forward-backward steps (until one does one optimizer step). But I can understand avoiding that because it makes the code more ugly.

**Comment by Gastron** [comment]
@RuABraun could you elaborate a little, I'm not sure if I understand. I think with this current scaling, with gradient accumulation factor `x`, the loss will be the sum of `x` batches' losses, each divided by `x`, which I think should keep the same scale as training without gradient accumulation. (I suppose this assumes loss reduction is mean-based or some such.)

**Comment by RuABraun** [comment]
You're write sorry! misread the code

**Comment by Gastron** [comment]
Phew, well thanks for double checking this!

File: speechbrain/core.py [hunk 1]
```diff
@@ -839,24 +852,29 @@ def fit_batch(self, batch):
...
             self.optimizer.zero_grad()
             with torch.cuda.amp.autocast():
                 outputs = self.compute_forward(batch, Stage.TRAIN)
                 loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-            self.scaler.scale(loss).backward()
-            self.scaler.unscale_(self.optimizer)
-            if self.check_gradients(loss):
-                self.scaler.step(self.optimizer)
-            self.scaler.update()
+            self.scaler.scale(loss/self.grad_accumulation_factor).backward()
```
Comment (danpovey) [inline speechbrain/core.py]: If you might be using DDP you might want to consider no_sync().
https://github.com/pytorch/pytorch/issues/43201

Comment (Gastron) [reply speechbrain/core.py]: Thanks, looks like that would speed things up! @TParcollet I think you've had some recipes using both gradient accumulation and DDP, got time to add this?
